features/lwipstack/LWIPStack.cpp

```diff
@@ -309,14 +309,27 @@ nsapi_error_t LWIP::socket_open(nsapi_socket_t *handle, nsapi_protocol_t proto)
         return NSAPI_ERROR_NO_SOCKET;
     }
 
-    enum netconn_type lwip_proto = proto == NSAPI_TCP ? NETCONN_TCP : NETCONN_UDP;
+    enum netconn_type netconntype;
+    if (proto == NSAPI_TCP) {
+        netconntype = NETCONN_TCP;
+    } else if (proto == NSAPI_UDP) {
+        netconntype = NETCONN_UDP;
+    } else {
+        netconntype = NETCONN_RAW;
+    }
+
 
 #if LWIP_IPV6
     // Enable IPv6 (or dual-stack)
-    lwip_proto = (enum netconn_type)(lwip_proto | NETCONN_TYPE_IPV6);
+    netconntype = (enum netconn_type)(netconntype | NETCONN_TYPE_IPV6);
 #endif
 
-    s->conn = netconn_new_with_callback(lwip_proto, &LWIP::socket_callback);
+    if (proto == NSAPI_RAW) {
+        s->conn = netconn_new_with_proto_and_callback(NETCONN_RAW,
+                                                      (u8_t)IP_PROTO_ICMP, &LWIP::socket_callback);
```

Comment:
ICMP being hard-coded here makes it an ICMP socket, not a raw socket.

Admittedly, ICMP is going to be the most common use case, so I'm wondering whether we should accept that, and just code it to be an ICMP socket (so rename the class and make it `NSAPI_ICMP`). Or whether we should make it possible to pass in the protocol here.

I'm not sure what off-board devices like ESP8266 and the cellular modems support? Do they support ICMP and/or raw IP?

Reply:
Yes I agree to re-naming this to ICMP Socket.